TrainingApp/src/screens/auth/sign-in/sign-in.js

```diff
@@ -0,0 +1,123 @@
+import { ROUTES } from '../../../navigation/routes';
+
+function onChangeEmailField(dispatch, text) {
+    if (text.includes('@')) {
+        dispatch(AuthActions.setSignInEmail({ email: text }));
+    } else {
+        dispatch(AuthActions.setSignInUserName({ userName: text }));
+    }
+}
+
+function onChangePasswordField(dispatch, password) {
+    dispatch(AuthActions.setSignInPassword({ password }));
+}
+
+function restorePassword({ navigation }) {
+    navigation.navigate(ROUTES.RESTORE_PASSWORD);
+}
+
+function requestSignIn(dispatch, { email, userName, password }) {
+    if ((email || userName) && password) {
+        dispatch(fetchSignIn({ email, userName, password }));
+    } else {
+        Vibration.vibrate(PATTERN.ONE);
+        Alert.alert('An Error Occurred!', "Username and password cannot be empty!", [{ text: 'Ok' }]);
+    }
+}
+
+function goToSignUp(navigation) {
+    navigation.navigate(ROUTES.SIGN_UP);
+}
+
+const SignIn = ({ navigation }) => {
+    const dispatch = useDispatch();
+
+    const { email, password, userName } = useSelector(state => state.auth.signInForm);
+    const { signInError } = useSelector(state => state.auth);
+
+    useEffect(() => {
+        if (signInError && signInError !== 'no_saved_token_found') {
+            Vibration.vibrate(PATTERN.ONE);
+            Alert.alert('An Error Occurred!', signInError, [{ text: 'Ok' }]);
+        }
+    }, [signInError]);
+
+    return (
+        <ScrollView>
+            <LinearGradient colors={BackgroundGradientColors} style={styles.container}>
+                <KeyboardAvoidingView behavior="position" style={styles.container}>
+                    <Text style={styles.title}>Ecommerce Store</Text>
+                    <TextInput
+                        value={email || userName}
+                        onChangeText={text => onChangeEmailField(dispatch, text)}
+                        style={styles.input}
+                        placeholder="Email Address or Login"
+                        placeholderTextColor={BaseStyles.colors.black}
+                        autoCompleteType="email"
+                        blurOnSubmit={true}
+                        importantForAutofill="yes"
+                        keyboardType="email-address"
+                        textContentType="emailAddress"
+                    />
+                    <TextInput
+                        value={password}
+                        onChangeText={updatePassword => onChangePasswordField(dispatch, updatePassword)}
+                        style={styles.input}
+                        placeholder="Password"
+                        placeholderTextColor={BaseStyles.colors.black}
+                        autoCompleteType="password"
+                        importantForAutofill="yes"
+                        secureTextEntry={true}
+                        textContentType="password"
+                    />
+                    <TouchableHighlight
```

Comment:
Always use a single component for all sign in , signup and login ,

Reply:
Using separate components a de-structured approach will be clean and easy to understand